components/locale/src/langid.rs

```diff
@@ -0,0 +1,151 @@
+use std::fmt::Write;
+use std::str::FromStr;
+
+use crate::parser::{parse_language_identifier, ParserError, ParserMode};
+use crate::subtags;
+
+/// `LanguageIdentifier` is a core struct representing a [`Unicode Language Identifier`].
+///
+/// # Example
+///
+/// ```
+/// use icu_locale::LanguageIdentifier;
+///
+/// let li: LanguageIdentifier = "en-US".parse()
+///     .expect("Failed to parse.");
+///
+/// assert_eq!(li.language, "en");
+/// assert_eq!(li.script, None);
+/// assert_eq!(li.region.unwrap(), "US");
+/// assert_eq!(li.variants.len(), 0);
+/// ```
+///
+/// # Parsing
+///
+/// Unicode recognizes three levels of standard conformance for any language identifier:
+///
+///  * *well-formed* - syntactically correct
+///  * *valid* - well-formed and only uses registered language subtags, extensions, keywords, types...
+///  * *canonical* - valid and no deprecated codes or structure.
+///
+/// At the moment parsing normalizes a well-formed language identifier converting
+/// `_` separators to `-` and adjusting casing to conform to the Unicode standard.
+///
+/// Any bogus subtags will cause the parsing to fail with an error.
+/// No subtag validation is performed.
+///
+/// # Example:
+///
+/// ```
+/// use icu_locale::LanguageIdentifier;
+///
+/// let li: LanguageIdentifier = "eN_latn_Us-Valencia".parse()
+///     .expect("Failed to parse.");
+///
+/// assert_eq!(li.language, "en");
+/// assert_eq!(li.script.unwrap(), "Latn");
+/// assert_eq!(li.region.unwrap(), "US");
+/// assert_eq!(li.variants.get(0).unwrap(), "valencia");
+/// ```
+///
+/// [`Unicode Language Identifier`]: https://unicode.org/reports/tr35/tr35.html#Unicode_language_identifier
+#[derive(Default, Debug, PartialEq, Eq, Clone, Hash, PartialOrd, Ord)]
+pub struct LanguageIdentifier {
+    /// Language subtag of the LanguageIdentifier
+    pub language: subtags::Language,
+    /// Script subtag of the LanguageIdentifier
+    pub script: Option<subtags::Script>,
+    /// Region subtag of the LanguageIdentifier
+    pub region: Option<subtags::Region>,
+    /// Variant subtags of the LanguageIdentifier
+    pub variants: subtags::Variants,
```

Comment:
Are we converting this in to plain 35 char string? Or some other opaque struct as we discussed in the meeting.

Reply:
For now, we're storing `Variants` which is `Box<[Variant]>` and each `Variant` is a `TinyStr8` as per spec.

In the follow up PR I'll propose remodeling that to treat `Variants` as a single `Box<[TinyStr8]>` just like [Transform Extension `tvalue`](https://unicode.org/reports/tr35/#transformed_extensions) is.